### PR TITLE
[YOLOV8] Return validation loss when running yolov8 `val.py`

### PR DESCRIPTION
**Pull request by dbogunowicz**
## Feature Preview
```bash
sparseml.ultralytics.val --model yolov8n.pt
```
Output:
```bash
Ultralytics YOLOv8.0.30 🚀 Python-3.8.10 torch-1.13.1+cu117 CUDA:0 (NVIDIA RTX A4000, 16117MiB)
yolo/engine/trainer: recipe=None, recipe_args=None, task=detect, mode=val, model=yolov8n.pt, data=coco.yaml, epochs=100, patience=50, batch=16, imgsz=640, save=True, save_period=-1, cache=False, device=None, workers=8, project=None, name=None, exist_ok=False, pretrained=False, optimizer=SGD, verbose=False, seed=0, deterministic=True, single_cls=False, image_weights=False, rect=True, cos_lr=False, close_mosaic=10, resume=False, min_memory=False, overlap_mask=True, mask_ratio=4, dropout=False, val=True, split=val, save_json=False, save_hybrid=False, conf=0.001, iou=0.7, max_det=300, half=False, dnn=False, plots=False, source=ultralytics/assets/, show=False, save_txt=False, save_conf=False, save_crop=False, hide_labels=False, hide_conf=False, vid_stride=1, line_thickness=3, visualize=False, augment=False, agnostic_nms=False, classes=None, retina_masks=False, boxes=True, format=torchscript, keras=False, optimize=False, int8=False, dynamic=False, simplify=False, opset=17, workspace=4, nms=False, lr0=0.01, lrf=0.01, momentum=0.937, weight_decay=0.001, warmup_epochs=3.0, warmup_momentum=0.8, warmup_bias_lr=0.1, box=7.5, cls=0.5, dfl=1.5, fl_gamma=0.0, label_smoothing=0.0, nbs=64, hsv_h=0.015, hsv_s=0.7, hsv_v=0.4, degrees=0.0, translate=0.1, scale=0.5, shear=0.0, perspective=0.0, flipud=0.0, fliplr=0.5, mosaic=1.0, mixup=0.0, copy_paste=0.0, cfg=None, v5loader=False, save_dir=runs/detect/val178
Ultralytics YOLOv8.0.30 🚀 Python-3.8.10 torch-1.13.1+cu117 CUDA:0 (NVIDIA RTX A4000, 16117MiB)
val: Scanning /home/ubuntu/damian/deepsparse/datasets/coco128/labels/train2017.cache... 126 images, 2 backgrounds, 0 corrupt: 100%|██████████| 128/128 [00:00<?, ?it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 8/8 [00:01<00:00,  4.05it/s]
                   all        128        929       0.64      0.537      0.605      0.446
Speed: 0.1ms pre-process, 1.9ms inference, 1.2ms loss, 0.6ms post-process per image
Validation loss: 1.07982
```
(results are the same as with the results in main)
### Testing train.py
Ran:
```bash
sparseml.ultralytics.train
```
all three losses (bbox, dfl and cls) go down consistently over dozens of epochs

### Testing eval on other task
Ran:
```bash
sparseml.ultralytics.val --model yolov8n.pt --task segment --data coco128-seg.yaml
```
Output:
```bash
Ultralytics YOLOv8.0.30 🚀 Python-3.8.10 torch-1.13.1+cu117 CUDA:0 (NVIDIA RTX A4000, 16117MiB)
yolo/engine/trainer: recipe=None, recipe_args=None, task=segment, mode=val, model=yolov8n-seg.pt, data=coco.yaml, epochs=100, patience=50, batch=16, imgsz=640, save=True, save_period=-1, cache=False, device=None, workers=8, project=None, name=None, exist_ok=False, pretrained=False, optimizer=SGD, verbose=False, seed=0, deterministic=True, single_cls=False, image_weights=False, rect=True, cos_lr=False, close_mosaic=10, resume=False, min_memory=False, overlap_mask=True, mask_ratio=4, dropout=False, val=True, split=val, save_json=False, save_hybrid=False, conf=0.001, iou=0.7, max_det=300, half=False, dnn=False, plots=False, source=ultralytics/assets/, show=False, save_txt=False, save_conf=False, save_crop=False, hide_labels=False, hide_conf=False, vid_stride=1, line_thickness=3, visualize=False, augment=False, agnostic_nms=False, classes=None, retina_masks=False, boxes=True, format=torchscript, keras=False, optimize=False, int8=False, dynamic=False, simplify=False, opset=17, workspace=4, nms=False, lr0=0.01, lrf=0.01, momentum=0.937, weight_decay=0.001, warmup_epochs=3.0, warmup_momentum=0.8, warmup_bias_lr=0.1, box=7.5, cls=0.5, dfl=1.5, fl_gamma=0.0, label_smoothing=0.0, nbs=64, hsv_h=0.015, hsv_s=0.7, hsv_v=0.4, degrees=0.0, translate=0.1, scale=0.5, shear=0.0, perspective=0.0, flipud=0.0, fliplr=0.5, mosaic=1.0, mixup=0.0, copy_paste=0.0, cfg=None, v5loader=False, save_dir=runs/segment/val19
Ultralytics YOLOv8.0.30 🚀 Python-3.8.10 torch-1.13.1+cu117 CUDA:0 (NVIDIA RTX A4000, 16117MiB)
val: Scanning /home/ubuntu/damian/deepsparse/datasets/coco128-seg/labels/train2017.cache... 126 images, 2 backgrounds, 0 corrupt: 100%|██████████| 128/128 [00:00<?, ?it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100%|██████████| 8/8 [00:02<00:00,  3.89it/s]
                   all        128        929      0.571      0.566      0.595      0.443      0.537      0.533      0.552      0.365
Speed: 0.1ms pre-process, 2.4ms inference, 2.3ms loss, 0.7ms post-process per image
Validation loss: 1.08198
```
(results are the same as with the results in main)

### CODE REVIEW

File: src/sparseml/yolov8/val.py [hunk 1]
```diff
@@ -29,7 +29,7 @@
 )
 @click.option(
     "--model",
-    default="yolov8n.yaml",
+    default="yolov8n.pt",
```
Comment (dbogunowicz) [inline src/sparseml/yolov8/val.py:32]:
```suggestion
    default="yolov8n.yaml",
```

File: src/sparseml/yolov8/validators.py [hunk 2]
```diff
@@ -30,9 +31,9 @@
 
 
 class SparseValidator(BaseValidator):
-    def __call__(self, trainer=None, model=None):
+    def __call__(self, trainer=None, model=None, training=True):
         # the **only** difference in this call is that we pass fuse=False to AutoBackend
```
Comment (corey-nm) [inline src/sparseml/yolov8/validators.py:35]: Can you remove this comment? No longer accurate and I think we have other changes planned for this function anyway

**Review by corey-nm — CHANGES_REQUESTED** [review]
I think this looks good. Can you also run a test to verify that the validation numbers before this PR and after this PR are identical? Want to make sure we aren't changing them somehow

File: src/sparseml/yolov8/validators.py [hunk 3]
```diff
@@ -121,8 +124,14 @@ def __call__(self, trainer=None, model=None):
 
             # loss
             with dt[2]:
-                if self.training:
-                    self.loss += trainer.criterion(preds, batch)[1]
+                if not hasattr(self, "loss"):
+                    self.loss = trainer.criterion(
```
[the remark this answers is not in the document]
Comment (corey-nm) [reply src/sparseml/yolov8/validators.py:127]: Oooooh gotcha, thanks for those details. This is fine then!

**Review by corey-nm — APPROVED** [review]
Thanks for the additional testing!
